### ISSUE TITLE
Having Default configuration, should'nt we give preference to router having higher interface IP ?

### ISSUE THREAD

**Issue by Navjot01** [acassen/keepalived#580]
In case we have two routers  {R1}(1.1.1.4) ---------------(1.1.1.3) {R2} both having same priority and all other configurations are default.

IMO, in such cases router having higher interface IP should become Master. But i don't see such behavior in the KA. Though i don't see such scenario in RFC but CISCO other routers have such behavior. It gives certain predictability and preference in case we have not set priority or any other parameters.

**Comment by pqarmitage** [comment]
You are correct that if the priorities are equal, then the router with the higher primary address to be the one to be master (unless nopreempt is configured). This is specified in RFC5798 in section 6.4.3 (735).

Unfortunately you don't say which version of keepalived you are using. Commit 9f67675 resolved the issue of not comparing addresses when receiving an equal priority advert, which was first included in version 1.2.22.

**Comment by Navjot01** [comment]
1.2.22 is the version which i am using, but commit shared by you is not present in my version i will include the same. Thanks !

**Comment by pqarmitage** [comment]
Commit 9f67675 was one of the last commits prior to the release of v1.2.22 (there were only two later commits, excluding merges), so if you don't have commit 9f67675 you don't really have v1.2.22.

You would probably be better off if you moved to a more recent version of keepalived, v1.3.5 is the current version.

**Comment by Navjot01** [comment]
Ok I will move try to move to newer version 1.3.5.

Thanks for the clarification.

**Comment by Navjot01** [comment]
Ok I will move try to move to newer version 1.3.5.

Thanks for the clarification.

**Comment by Navjot01** [comment]
I have checked the patch and its already present in my code base, i also verified that version which i am using is 1.2.22. Still i am seeing the election problem. After debugging i found that default state for instance on creation i am using is BACKUP i.s.

[R1] (1.1.1.2)------------- (1.1.1.1)[R2], here i have created R2 first & it is acting as Master now when i create instance on R1 it comes up with state BACKUP because that is the state i am using as default, so even though interface IP of R1 is greater as compare to R2 still R2 is acting as Master.

What should be the default state for new instance ?

We kept it as BACKUP to make sure that new router does not start transferring the packets it should wait and if required change its state to MASTER.

Can you please share your inputs on this scenario ?

**Comment by pqarmitage** [comment]
Could you please provide a copy of your configuration so that I can try it and see what is happening.

**Comment by Navjot01** [comment]
Keeping in mind above R1 & R2 example, first i configured on R2 below is the configuration as you can see default state i am using over here is BACKUP. As other side instance is not configured so R2 will become Master.
vrrp_instance vrrp_1 {
	state              BACKUP
	interface  swid0_eth.11
	virtual_router_id 1
	use_vmac
	priority       100
	advert_int 1 
	garp_master_delay 1 
	preempt
	authentication {
		auth_type NONE

	}
	virtual_ipaddress {
	 1.1.1.5
       }
	notify_master "/etc/keepalived/notify_master.sh 1"
	notify_backup "/etc/keepalived/notify_backup.sh 1"
	notify_fault  "/etc/keepalived/notify_fault.sh 1"
}

Now i configure below on R1 with same default state i.s. BACKUP, once this instance is configured it will start receiving the packets from R2 and state will remain in Backup, even though interface address over here is higher.
vrrp_instance vrrp_1 {
	state              BACKUP
	interface  swid0_eth.11
	virtual_router_id 1
	use_vmac
	priority       100
	advert_int 1 
	garp_master_delay 1 
	preempt
	authentication {
		auth_type NONE

	}
	virtual_ipaddress {
	 1.1.1.5
       }
	notify_master "/etc/keepalived/notify_master.sh 1"
	notify_backup "/etc/keepalived/notify_backup.sh 1"
	notify_fault  "/etc/keepalived/notify_fault.sh 1"
}

But if i keep default state as MASTER on new instance creation i see correct router as Master, this is obvious because we have this logic in master_rx not in vrrp_state_backup.

I hope this will help you to  understand the scenario. Let me know in case of any other input. IMO we should also add address check in vrrp_state_backup so that it can change its state to Master in case of default configuration.

**Comment by pqarmitage** [comment]
Having tried the above, I realise that what you are seeing is the correct behaviour.

In RFC5798, at (445) in section 6.4.2, if a backup receives an advert of greater than or equal priority to its own priority, then it simply resets its timer and carries on. At (465), which is where a backup will start sending adverts, is only of the advert is of LOWER priority.

The point where addresses are compared when priorities are equal is if a MASTER receives an advert (735), and the address comparison is used to break a tie when two (or more) systems are both/all sending adverts.

In summary, a backup router will not preempt unless it sees adverts with a strictly lower priority, which explains why the second router starting up will never take over from the first when the priorities are equal.

**Comment by Navjot01** [comment]
I totally agree with your explanation, point is what should be the default value of state when we create new instance.

Because if i set the instance default state as backup, election does not take place which is aligned with your above explanation, in case i set default state as Master, election takes place which is also as per the above explanation.  But do we have any specific recommendation to use backup or master as default value for state or we should keep it as implementation dependent ?

**Comment by pqarmitage** [comment]
My recommendation would be to set state to BACKUP (or don't configure it at all). Setting it to MASTER is causing unexpected behaviour.

BTW, specifying auth_type NONE is not a valid auth type; simply remove the authentication block.
